File: c/medium/minimum_ascii_delete_sum_for_two_strings.c
#include <stdio.h>
#include <string.h>

int max(int a, int b) {
    return a > b ? a : b;
}

int minimumDeleteSum(char* s1, char* s2) {
    int len_a = strlen(s1);
    int len_b = strlen(s2);

    // Ensure len_a >= len_b
    if (len_a < len_b) {
        return minimumDeleteSum(s2, s1);
    }

    int dp[2][1001] = {0}; // assuming max length <= 1000

    for (int i = 0; i < len_a; i++) {
        int curr = (i + 1) & 1;
        int prev = curr ^ 1;

        for (int j = 0; j < len_b; j++) {
            if (s1[i] == s2[j]) {
                dp[curr][j + 1] = (int)s1[i] + dp[prev][j];
            } else {
                dp[curr][j + 1] = max(dp[prev][j + 1], dp[curr][j]);
            }
        }
    }

    int total = 0;
    for (int i = 0; i < len_a; i++) total += (int)s1[i];
    for (int j = 0; j < len_b; j++) total += (int)s2[j];

    int max_common = dp[len_a & 1][len_b];
    return total - 2 * max_common;
}

File: rust/medium/minimum_ascii_delete_sum_for_two_strings.rs
pub fn minimum_delete_sum(s1: String, s2: String) -> i32 {
    if s1.len() < s2.len() {
        return minimum_delete_sum(s2, s1);
    }

    let a = s1.as_bytes();
    let b = s2.as_bytes();

    let len_a = a.len();
    let len_b = b.len();

    let mut dp = vec![vec![0i32; len_b + 1]; 2];

    for i in 0..len_a {
        let curr = (i + 1) & 1;
        let prev = curr ^ 1;

        for j in 0..len_b {
            if a[i] == b[j] {
                dp[curr][j + 1] = a[i] as i32 + dp[prev][j];
            } else {
                dp[curr][j + 1] = dp[prev][j + 1].max(dp[curr][j]);
            }
        }
    }

    let total: i32 = a.iter().map(|&c| c as i32).sum::<i32>()
        + b.iter().map(|&c| c as i32).sum::<i32>();

    let max_common = dp[len_a & 1][len_b];
    total - 2 * max_common
}
